Errors during scheduled flush of TransformationCache

Recently we saw errors in our fc logs that might indicate a problem.

```
PreparedStatement can have at most 65,535 parameters. Please consider using arrays, or splitting the query in several ones, or using COPY. Given query has 83,791 parameters
	at org.springframework.jdbc.support.SQLStateSQLExceptionTranslator.doTranslate(SQLStateSQLExceptionTranslator.java:104)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:70)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:79)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:79)
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1541)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:667)
	at org.springframework.jdbc.core.JdbcTemplate.update(JdbcTemplate.java:960)
	at org.springframework.jdbc.core.JdbcTemplate.update(JdbcTemplate.java:981)
	at org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate.update(NamedParameterJdbcTemplate.java:328)
	at org.factcast.store.registry.transformation.cache.PgTransformationCache.insertBufferedAccesses(PgTransformationCache.java:214)
	at org.factcast.store.registry.transformation.cache.PgTransformationCache.flush(PgTransformationCache.java:176)
	at jdk.internal.reflect.GeneratedMethodAccessor172.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:84)
	at io.opentelemetry.javaagent.instrumentation.spring.scheduling.SpringSchedulingRunnableWrapper.run(SpringSchedulingRunnableWrapper.java:35)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.postgresql.util.PSQLException: PreparedStatement can have at most 65,535 parameters. Please consider using arrays, or splitting the query in several ones, or using COPY. Given query has 83,791 parameters
	at org.postgresql.jdbc.PgPreparedStatement.<init>(PgPreparedStatement.java:102)
	at org.postgresql.jdbc.PgPreparedStatement.<init>(PgPreparedStatement.java:88)
	at org.postgresql.jdbc.PgConnection.prepareStatement(PgConnection.java:1325)
	at org.postgresql.jdbc.PgConnection.prepareStatement(PgConnection.java:1784)
	at org.postgresql.jdbc.PgConnection.prepareStatement(PgConnection.java:454)
	at jdk.internal.reflect.GeneratedMethodAccessor96.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.apache.tomcat.jdbc.pool.ProxyConnection.invoke(ProxyConnection.java:131)
	at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:109)
	at org.apache.tomcat.jdbc.pool.interceptor.AbstractCreateStatementInterceptor.invoke(AbstractCreateStatementInterceptor.java:75)
	at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:109)
	at org.apache.tomcat.jdbc.pool.DisposableConnectionFacade.invoke(DisposableConnectionFacade.java:81)
	at jdk.proxy3/jdk.proxy3.$Proxy137.prepareStatement(Unknown Source)
	at org.springframework.jdbc.core.PreparedStatementCreatorFactory$PreparedStatementCreatorImpl.createPreparedStatement(PreparedStatementCreatorFactory.java:235)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:649)
	... 17 common frames omitted

```

What I get from the code, I would say this shouldn't be possible at all because theres an upper limit

--- a/factcast-store/src/main/java/org/factcast/store/registry/transformation/cache/PgTransformationCache.java
+++ b/factcast-store/src/main/java/org/factcast/store/registry/transformation/cache/PgTransformationCache.java
@@ -41,7 +41,7 @@ import org.springframework.scheduling.annotation.Scheduled;
 @RequiredArgsConstructor
 @Slf4j
 public class PgTransformationCache implements TransformationCache {
-  public static final int MAX_BATCH_SIZE = 32_000;
+  private static final int MAX_BATCH_SIZE = 20_000;
   private final JdbcTemplate jdbcTemplate;
   private final NamedParameterJdbcTemplate namedJdbcTemplate;
 

--- a/factcast-store/src/main/java/org/factcast/store/registry/transformation/cache/PgTransformationCache.java
+++ b/factcast-store/src/main/java/org/factcast/store/registry/transformation/cache/PgTransformationCache.java
@@ -41,7 +41,7 @@ import org.springframework.scheduling.annotation.Scheduled;
 @RequiredArgsConstructor
 @Slf4j
 public class PgTransformationCache implements TransformationCache {
-  public static final int MAX_BATCH_SIZE = 10000;
+  public static final int MAX_BATCH_SIZE = 32_000;
   private final JdbcTemplate jdbcTemplate;
   private final NamedParameterJdbcTemplate namedJdbcTemplate;
 

--- a/factcast-store/src/main/java/org/factcast/store/registry/transformation/cache/PgTransformationCache.java
+++ b/factcast-store/src/main/java/org/factcast/store/registry/transformation/cache/PgTransformationCache.java
@@ -16,6 +16,7 @@
 package org.factcast.store.registry.transformation.cache;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Iterables;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import java.time.ZonedDateTime;
@@ -40,6 +41,7 @@ import org.springframework.scheduling.annotation.Scheduled;
 @RequiredArgsConstructor
 @Slf4j
 public class PgTransformationCache implements TransformationCache {
+  public static final int MAX_BATCH_SIZE = 10000;
   private final JdbcTemplate jdbcTemplate;
   private final NamedParameterJdbcTemplate namedJdbcTemplate;
 
@@ -200,11 +202,15 @@ public class PgTransformationCache implements TransformationCache {
             .collect(Collectors.toList());
 
     if (!parameters.isEmpty()) {
+
       // dup-keys can be ignored, in case another node just did the same
-      jdbcTemplate.batchUpdate(
-          "INSERT INTO transformationcache (cache_key, header, payload) VALUES (?, ? :: JSONB, ? ::"
-              + " JSONB) ON CONFLICT(cache_key) DO NOTHING",
-          parameters);
+      Iterables.partition(parameters, MAX_BATCH_SIZE)
+          .forEach(
+              p ->
+                  jdbcTemplate.batchUpdate(
+                      "INSERT INTO transformationcache (cache_key, header, payload) VALUES (?, ? :: JSONB, ? ::"
+                          + " JSONB) ON CONFLICT(cache_key) DO NOTHING",
+                      p));
     }
   }
 
@@ -217,9 +223,15 @@ public class PgTransformationCache implements TransformationCache {
             .collect(Collectors.toList());
 
     if (!keys.isEmpty()) {
-      SqlParameterSource parameters = new MapSqlParameterSource("ids", keys);
-      namedJdbcTemplate.update(
-          "UPDATE transformationcache SET last_access=now() WHERE cache_key IN (:ids)", parameters);
+
+      Iterables.partition(keys, MAX_BATCH_SIZE)
+          .forEach(
+              k -> {
+                SqlParameterSource parameters = new MapSqlParameterSource("ids", k);
+                namedJdbcTemplate.update(
+                    "UPDATE transformationcache SET last_access=now() WHERE cache_key IN (:ids)",
+                    parameters);
+              });
     }
   }
 }

--- a/factcast-store/src/main/java/org/factcast/store/registry/transformation/cache/PgTransformationCache.java
+++ b/factcast-store/src/main/java/org/factcast/store/registry/transformation/cache/PgTransformationCache.java
@@ -130,8 +130,15 @@ public class PgTransformationCache implements TransformationCache {
     registryMetrics.increase(EVENT.TRANSFORMATION_CACHE_MISS, misses);
     registryMetrics.increase(EVENT.TRANSFORMATION_CACHE_HIT, hits);
 
-    buffer.putAllNull(keys);
+    registerAccess(keys);
+
     return Sets.newHashSet(facts);
+  }
+
+  @VisibleForTesting
+  CompletableFuture<Void> registerAccess(Collection<Key> keys) {
+    buffer.putAllNull(keys);
+    return flushIfNecessary();
   }
 
   @VisibleForTesting

--- a/factcast-store/src/test/java/org/factcast/store/registry/transformation/cache/PgTransformationCacheTest.java
+++ b/factcast-store/src/test/java/org/factcast/store/registry/transformation/cache/PgTransformationCacheTest.java
@@ -302,6 +302,16 @@ class PgTransformationCacheTest {
     }
 
     @Test
+    @SneakyThrows
+    void afterAcess_list() {
+      underTest.registerAccess(List.of(key, key2)).get();
+
+      assertThat(underTest.buffer().size()).isEqualTo(2);
+      assertThat(underTest.buffer().buffer().values()).allMatch(x -> x == null);
+      verify(underTest).flushIfNecessary();
+    }
+
+    @Test
     void logsException() {
       underTest.registerAccess(key2);
       underTest.registerWrite(key, f);

--- a/factcast-store/src/test/java/org/factcast/store/registry/transformation/cache/PgTransformationCacheTest.java
+++ b/factcast-store/src/test/java/org/factcast/store/registry/transformation/cache/PgTransformationCacheTest.java
@@ -154,17 +154,22 @@ class PgTransformationCacheTest {
       // only one key is looked for in persistent cache
       Collection ids = (Collection) cap.getValue().getValue("ids");
       assertThat(ids).hasSize(1).doesNotContain(key);
+
+      verify(underTest).registerAccess(Lists.newArrayList(key2));
     }
 
     @Test
     void findsAllInCache() {
+      final var keysToFind = Lists.newArrayList(key, key2);
+
       Mockito.when(
               namedJdbcTemplate.query(
                   anyString(), any(SqlParameterSource.class), any(RowMapper.class)))
           .thenReturn(Lists.newArrayList(f, f2));
-      assertThat(underTest.findAll(Lists.newArrayList(key, key2)))
-          .hasSize(2)
-          .containsExactlyInAnyOrder(f, f2);
+
+      assertThat(underTest.findAll(keysToFind)).hasSize(2).containsExactlyInAnyOrder(f, f2);
+
+      verify(underTest).registerAccess(keysToFind);
     }
   }
 

